tests/nn/data_parallel/test_fsdp.py

```diff
@@ -305,6 +314,24 @@ def test_cpu_offload_and_cpu_grads(self):
         )
         spawn_and_init(test_fn)
 
+    def test_no_cpu_offload_with_sharded_grad_scaler(self):
+        # We don't test the False condition because that requires the optimizer to internally do
+        # the device transfer and PyTorch optimizers don't support this.
+        config = {"mixed_precision": False, "cpu_offload": False, "move_grads_to_cpu": False}
+        test_fn = functools.partial(
+            self._test_identical_outputs, TransformerWithSharedParams, config, use_cuda=True, lr=0.01
+        )
+        spawn_and_init(test_fn)
+
+    def test_no_cpu_offload_with_sharded_grad_scaler_and_mixed_precision(self):
```

Comment:
Can we use parametrization to combine these 4 tests? you are testing mixed_precision=True/False and cpu_offload=True/False. we can skip the move_grads_to_cpu since that is set by default by the cpu_offload param.  Also wanted to mention that we have deprecated the cpu_offload param and should use `move_params_to_cpu` instead.

Reply:
done.

Reply:
Upon changing the `config` dictionary key from "cpu_offload" to "move_params_to_cpu" the `test_cpu_offload_and_cpu_grads` test breaks with the following error message:

```E       -- Process 0 terminated with the following error:
E       Traceback (most recent call last):
E         File "/private/home/anupamb/fairscale/tests/nn/data_parallel/test_fsdp.py", line 133, in _test_identical_outputs
E           torch.testing.assert_allclose(ref_loss, shard_loss)
E         File "/private/home/anupamb/miniconda3/lib/python3.9/site-packages/torch/testing/__init__.py", line 222, in assert_allclose
E           result, debug_msg = _compare_tensors_internal(actual, expected,
E         File "/private/home/anupamb/miniconda3/lib/python3.9/site-packages/torch/testing/__init__.py", line 130, in _compare_tensors_internal
E           if torch.allclose(a, b, rtol=rtol, atol=atol, equal_nan=equal_nan):
E       RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cpu!
E
E       During handling of the above exception, another exception occurred:
E
E       Traceback (most recent call last):
E         File "/private/home/anupamb/miniconda3/lib/python3.9/site-packages/torch/multiprocessing/spawn.py", line 59, in _wrap
E           fn(i, *args)
E         File "/private/home/anupamb/fairscale/tests/nn/data_parallel/test_fsdp.py", line 814, in init_and_run
E           fn(rank, group, *args)
E         File "/private/home/anupamb/fairscale/tests/nn/data_parallel/test_fsdp.py", line 136, in _test_identical_outputs
E           raise Exception(f"FullyShardedDataParallel didn't match PyTorch DDP using config: {config}\n\n {e}")
E       Exception: FullyShardedDataParallel didn't match PyTorch DDP using config: {'mixed_precision': True, 'move_params_to_cpu': True, 'compute_dtype': torch.float32}
E
E        Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cpu!
```